### PR TITLE
All the JS changes.

### PR DESCRIPTION
**Pull request by namin**
Hi Tiark,

This is all the JS changes on top of delite-develop. Would you consider incorporating them? I can split this into multiple commits or rebase if it helps.

Thanks.

### CODE REVIEW

File: src/common/Functions.scala [hunk 1]
```diff
@@ -18,62 +20,91 @@ trait Functions extends Base {
...
+  implicit def toLambdaOps3[A1:Manifest,A2:Manifest,A3:Manifest,B:Manifest](fun: Rep[((A1,A2,A3)) => B]) =
+    new LambdaOps3(fun)
+  implicit def toLambdaOps4[A1:Manifest,A2:Manifest,A3:Manifest,A4:Manifest,B:Manifest](fun: Rep[((A1,A2,A3,A4)) => B]) =
+    new LambdaOps4(fun)
+  implicit def toLambdaOps5[A1:Manifest,A2:Manifest,A3:Manifest,A4:Manifest,A5:Manifest,B:Manifest](fun: Rep[((A1,A2,A3,A4,A5)) => B]) =
+    new LambdaOps5(fun)
 }
 
 trait FunctionsExp extends Functions with EffectExp {
+  case class Lambda[A:Manifest,B:Manifest](f: Exp[A] => Exp[B], x: Exp[A], y: Block[B]) extends Def[A => B]
```
Comment (TiarkRompf) [inline src/common/Functions.scala:66]: why is x an Exp and not a Sym?

File: src/common/Functions.scala [hunk 2]
```diff
@@ -84,24 +115,128 @@ trait FunctionsExp extends Functions with EffectExp {
...
+      UnboxedTuple[A](mA.typeArguments.map(fresh(_)))
+    else fresh[A]
+  }
+
+  override def unbox[A:Manifest](x : Exp[A])(implicit pos: SourceContext) : Exp[A] = {
+    val mA = implicitly[Manifest[A]]
+    x match {
+      case _ : UnboxedTuple[A] => x
+      case _ if mA == implicitly[Manifest[Unit]] =>
+	UnboxedTuple[A](List())
```
Comment (TiarkRompf) [inline src/common/Functions.scala:149]: tabs vs spaces?

**Comment by TiarkRompf** [comment]
Thanks for putting this together. Do all tests pass? If so I'll merge it in right away.

**Comment by namin** [comment]
Cool. All the tests pass, except for the two code motion tests that were already broken.

File: src/common/Functions.scala (hunk 1 above)
Comment (namin) [inline src/common/Functions.scala:66]: Because sometimes it can be a tuple.

File: src/common/Functions.scala (hunk 2 above)
Comment (namin) [inline src/common/Functions.scala:149]: Done in new pull request.

**Pull request merged**